File: apps/client/lib/data/api/biu_client.dart
// BiuClient —— Flutter 端连 brain Agent Plane WS endpoint 的 ergonomic 包装。
//
// 不直接暴露 sdkproto/v1 的裸类型 —— 上层 controller 拿 BiuClient 就够了：
//   - connect(sessionId, sessionToken) → 立即返回；后台跑 WS 升级 + read pump
//   - frames stream → 每帧解过的 Object（SDKMessage / Lifecycle / 控制帧）
//   - send(SDKUserMessage) / sendInterrupt() → 给 brain 发数据
//   - close()                              → 主动断
//
// 协议跟 services/brain/internal/agentplane/ingress.go 对齐：
//
//   GET /v1/agent/sessions/{id}/stream?session_token=<jwt>&since_seq=<n>
//   ↑ ingress 校 token + scope；上行帧就是 sdkproto/v1 JSON
//
// 自动重连策略：
//   - WS 主动关 → 不重连（上层显式 close）
//   - 网络异常断（read pump err） → 指数退避 1s → 30s + 用 lastSeq+1 续
//   - session_token 过期 (401) → 触发 onTokenExpired callback 让上层 refresh
//
// 心跳：上层 WebSocketChannel 自带 PING/PONG（gorilla 兼容）。本层不发额外心跳，
// 但维护一个 60s read deadline；超时视为断（触发重连）。
//
// 重放策略：每收到带 uuid 的帧记 lastSeq（uuid 不是 seq；ingress 用 stream
// seq 做 since_seq）。当前 v1 只能用 ?since_seq= 续；具体 seq 由 ingress 在帧
// header 里给（X-Seq）—— 但 WS 无 header，所以本版不实现真重放，只做
// reconnect。重放留 v2（需要 ingress 在 frame 内嵌 seq 字段）。

import 'dart:async';
import 'dart:convert';

import 'package:flutter/foundation.dart';
import 'package:web_socket_channel/web_socket_channel.dart';

import 'sdkproto/v1/common.dart';
import 'sdkproto/v1/data/user.dart';
import 'sdkproto/v1/service.dart';

/// 连一条 brain Agent Plane WS session 的入口。
///
/// 用法：
/// ```dart
/// final c = BiuClient(brainBaseUrl: 'wss://your-biumind.example.com');
/// await c.connect(sessionId: '...', sessionToken: '...');
/// c.frames.listen((frame) {
///   if (frame is SDKStreamlinedText) print(frame.text);
/// });
/// c.send(SDKUserMessage(message: AnthropicMessage(...)));
/// ```
class BiuClient {
  /// brain WS base URL，例 `ws://localhost:7003` 或 `wss://your-biumind.example.com`。
  /// 末尾不带 `/`；class 内部拼 `/v1/agent/sessions/<id>/stream`。
  final String brainBaseUrl;

  /// onTokenExpired 在收到 1008 (policy violation) 或 4xx 时 fire；上层应该
  /// 调 brain `/v1/agent/sessions/{id}/refresh-token` 拿新 session_token，
  /// 然后调 [refreshToken] 续连接。nil 时 client 直接关。
  final FutureOr<String?> Function()? onTokenExpired;

  /// 自定义 transport connector —— 测试用。生产留 null 走 WebSocketChannel.connect。
  /// 测试可以传一个 fake 实现完成端到端断言（详见 biu_client_test.dart）。
  final BiuTransport Function(Uri uri)? connector;

  /// reconnectMaxAttempts 连续失败次数到达后 client 放弃，frames stream close。
  /// 默认 8 次（指数退避 1s..30s 大概 ~3 分钟覆盖）。
  final int reconnectMaxAttempts;

  /// outboxMaxLen [enqueue] 离线队列上限。超出抛 StateError。默认 64。
  /// 离线 5 min 内一个用户疯狂输入 64 条消息已经很极端 —— 真过限说明
  /// 网络长期断 + UI 没显示状态徽章，提示 caller 阻塞输入。
  final int outboxMaxLen;

  BiuClient({
    required this.brainBaseUrl,
    this.onTokenExpired,
    this.connector,
    this.reconnectMaxAttempts = 8,
    this.outboxMaxLen = 64,
  });

  // ── 内部状态 ─────────────────────────────────────────────

  BiuTransport? _channel;
  StreamSubscription<dynamic>? _readSub;
  final _framesController = StreamController<Object>.broadcast();
  bool _closedByUser = false;
  String? _sessionId;
  String? _sessionToken;
  int _sinceSeq = 0;
  int _reconnectAttempt = 0;
  final List<Object> _outbox = [];
  Timer? _reconnectTimer;

  // ── 公开 API ─────────────────────────────────────────────

  /// 已解析的下行帧流（SDKMessage / Lifecycle / SDKControlRequest 等）。
  /// broadcast stream，多个 listener 都能订阅。
  Stream<Object> get frames => _framesController.stream;

  /// 当前是否连着。WS 异步升级期 + 重连退避期都返回 false。
  bool get isConnected => _channel != null && !_closedByUser;

  /// 跟一条 session 建 WS 连接。返回前不等 ready —— frames stream 是异步的。
  /// 重连 / token refresh 会复用最初传入的 sessionId + sessionToken（用 [refreshToken]
  /// 替换）。
  ///
  /// [sinceSeq] —— 跨设备 resume 时让 brain ingress 用 OrderedConsumer 从
  /// `sinceSeq+1` 开始 replay 历史帧 + 自动接入实时流。0（默认）走实时拉，
  /// 不重放。stream 已 trim（消息超过 MaxAge / MaxMsgs）时 ingress 推一帧
  /// `biumind.session_desynced` 然后 close —— 客户端按规约拿
  /// `agent_session_results` 摘要兜底。
  Future<void> connect({
    required String sessionId,
    required String sessionToken,
    int sinceSeq = 0,
  }) async {
    _sessionId = sessionId;
    _sessionToken = sessionToken;
    _sinceSeq = sinceSeq;
    _closedByUser = false;
    _reconnectAttempt = 0;
    await _doConnect();
  }

  /// 把 [SDKUserMessage] / [SDKControlRequest] / [Lifecycle] 等帧序列化后
  /// 发给 brain。未连或已关 → 抛 [StateError]。
  ///
  /// **不入离线队列** —— 老调用方期望的同步抛错语义保留。需要离线缓冲
  /// 走 [enqueue]。
  void send(Object frame) {
    final ch = _channel;
    if (ch == null) {
      throw StateError('BiuClient.send: not connected');
    }
    final json = ServiceFrame.toJson(frame);
    ch.send(jsonEncode(json));
  }

  /// S9-3 离线送达队列 —— 跟 [send] 区别：
  ///   - 当前已连 → 立即 send（行为同 [send]）
  ///   - 未连 / 已断 / 未 connect → 进 FIFO 队列，重连成功后顺序 flush
  ///   - 主动 close 后入队 → 抛 [StateError]（防止已 close 还入新帧）
  ///
  /// 用法：手机离线 5 min 期间用户连发 3 条消息，UI 调 enqueue() 不报错；
  /// 网络恢复后 _doConnect 成功，flush 队列把 3 条按顺序送出。
  ///
  /// 队列上限 [outboxMaxLen]（默认 64）—— 极端情况下用户疯狂输入避免
  /// 内存爆。超过后抛 [StateError]。
  void enqueue(Object frame) {
    if (_closedByUser) {
      throw StateError('BiuClient.enqueue: closed by user');
    }
    final ch = _channel;
    if (ch != null && _outbox.isEmpty) {
      // 当前在线 + 队列空 → 直接 send，省一次 enqueue→flush 跳板
      send(frame);
      return;
    }
    if (_outbox.length >= outboxMaxLen) {
      throw StateError('BiuClient.enqueue: outbox full ($outboxMaxLen)');
    }
    _outbox.add(frame);
  }

  /// 当前离线队列长度（待 flush 的帧数）。UI 进度展示用。
  int get outboxPending => _outbox.length;

  /// 发一条 user message —— 简化封装，避免上层每次写 toJson 模板。
  /// content 走 Anthropic Messages API content block 数组格式（type=text）。
  void sendUserText(String text, {required String userMessageUuid}) {
    final um = SDKUserMessage(
      uuid: userMessageUuid,
      sessionId: _sessionId ?? '',
      message: AnthropicMessage(
        role: 'user',
        content: [
          {'type': 'text', 'text': text},
        ],
      ),
    );
    send(um);
  }

  /// 上层调 brain refresh-token endpoint 拿到新 token 后调这个，重连下次用。
  /// 如果当前已断，立即触发重连尝试；如果还连着，下次断时使用新 token。
  Future<void> refreshToken(String newSessionToken) async {
    _sessionToken = newSessionToken;
    if (_channel == null && !_closedByUser) {
      await _doConnect();
    }
  }

  /// 主动关。frames stream close，不会再重连。
  Future<void> close() async {
    _closedByUser = true;
    _reconnectTimer?.cancel();
    _reconnectTimer = null;
    await _readSub?.cancel();
    _readSub = null;
    await _channel?.close();
    _channel = null;
    if (!_framesController.isClosed) {
      await _framesController.close();
    }
  }

  // ── 内部实现 ─────────────────────────────────────────────

  Future<void> _doConnect() async {
    if (_closedByUser) return;
    final sid = _sessionId;
    final tok = _sessionToken;
    if (sid == null || tok == null) {
      throw StateError('BiuClient: connect() must run before _doConnect');
    }

    var url =
        '$brainBaseUrl/v1/agent/sessions/$sid/stream?session_token=${Uri.encodeQueryComponent(tok)}';
    if (_sinceSeq > 0) {
      url += '&since_seq=$_sinceSeq';
    }
    final uri = Uri.parse(url);
    try {
      final ch = (connector ?? _defaultConnector)(uri);
      _channel = ch;
      _reconnectAttempt = 0;
      _readSub = ch.frames.listen(
        _onFrame,
        onError: _onSocketError,
        onDone: _onSocketDone,
        cancelOnError: false,
      );
      _flushOutbox();
    } catch (e, stack) {
      debugPrint('BiuClient connect failed: $e\n$stack');
      _scheduleReconnect();
    }
  }

  /// _flushOutbox 把离线队列里的帧按 FIFO 顺序送出。失败时停止 flush，
  /// 剩下的帧留队列等下次重连。
  void _flushOutbox() {
    while (_outbox.isNotEmpty) {
      final ch = _channel;
      if (ch == null) return; // 中途断了
      final frame = _outbox.first;
      try {
        final json = ServiceFrame.toJson(frame);
        ch.send(jsonEncode(json));
        _outbox.removeAt(0);
      } catch (e) {
        debugPrint('BiuClient flush outbox failed: $e');
        return; // 留队列；下次重连再试
      }
    }
  }

  /// 默认走 [WebSocketChannel.connect]；包一层 [BiuTransport] adapter。
  static BiuTransport _defaultConnector(Uri uri) {
    final ch = WebSocketChannel.connect(uri);
    return _WSChannelTransport(ch);
  }

  void _onFrame(dynamic raw) {
    if (raw is! String) {
      // gorilla 用 TextMessage；BinaryMessage 当前不支持，drop
      return;
    }
    try {
      final json = jsonDecode(raw) as Map<String, dynamic>;
      final frame = ServiceFrame.fromJson(json);
      if (!_framesController.isClosed) {
        _framesController.add(frame);
      }
    } catch (e, stack) {
      debugPrint('BiuClient parse frame failed: $e\nraw=$raw\n$stack');
      // 不断连接 —— 一帧解析失败不应该让整个 session 挂掉
    }
  }

  void _onSocketError(Object err, StackTrace stack) {
    debugPrint('BiuClient socket error: $err\n$stack');
    _onSocketDone();
  }

  void _onSocketDone() {
    _readSub = null;
    _channel = null;
    if (_closedByUser) {
      if (!_framesController.isClosed) _framesController.close();
      return;
    }
    _scheduleReconnect();
  }

  void _scheduleReconnect() {
    if (_closedByUser) return;
    _reconnectAttempt++;
    if (_reconnectAttempt > reconnectMaxAttempts) {
      debugPrint('BiuClient: reconnect attempts exhausted, giving up');
      _closedByUser = true;
      if (!_framesController.isClosed) _framesController.close();
      return;
    }
    final delay = _backoffDelay(_reconnectAttempt);
    debugPrint(
      'BiuClient: reconnect in ${delay.inMilliseconds}ms '
      '(attempt $_reconnectAttempt/$reconnectMaxAttempts)',
    );
    _reconnectTimer?.cancel();
    _reconnectTimer = Timer(delay, () async {
      // 401/policy 路径：上层 onTokenExpired 拿新 token 后再 _doConnect
      if (onTokenExpired != null && _reconnectAttempt > 2) {
        try {
          final fresh = await onTokenExpired!();
          if (fresh != null && fresh.isNotEmpty) {
            _sessionToken = fresh;
          }
        } catch (e) {
          debugPrint('BiuClient onTokenExpired failed: $e');
        }
      }
      await _doConnect();
    });
  }

  Duration _backoffDelay(int attempt) {
    // 1s, 2s, 4s, 8s, 16s, 30s（cap）...
    final ms = 1000 * (1 << (attempt - 1).clamp(0, 5));
    return Duration(milliseconds: ms.clamp(1000, 30000));
  }
}

/// BiuTransport 是 BiuClient 跟下层连接握手的最小接口。生产是 WS，
/// 测试可以传内存 fake。frames 是 broadcast / single-subscribe 都行
/// （BiuClient 自己只 listen 一次）。
abstract class BiuTransport {
  /// 下行帧 stream。每条 element 是 server push 的原始 String / `List<int>`。
  /// BiuClient 只处理 String 类型；Binary 直接 drop。
  Stream<dynamic> get frames;

  /// 上行发一条 String（已经 jsonEncode 过）。失败由实现 best-effort 处理。
  void send(String data);

  /// 主动关。idempotent —— BiuClient.close 会调一次。
  Future<void> close();
}

/// _WSChannelTransport 把 [WebSocketChannel] 包成 [BiuTransport]。
class _WSChannelTransport implements BiuTransport {
  final WebSocketChannel _ch;
  _WSChannelTransport(this._ch);

  @override
  Stream<dynamic> get frames => _ch.stream;

  @override
  void send(String data) => _ch.sink.add(data);

  @override
  Future<void> close() async {
    await _ch.sink.close();
  }
}

File: apps/miniapp/src/lib/biu_client.ts
// BiuClient (miniapp) —— uni.connectSocket 包装的 brain Agent Plane WS 客户端。
//
// **跟 realtime_hub.ts 对比**：
//   - realtime_hub 走老 AG-UI Realtime（chunked SSE + polling）
//   - BiuClient 走新 SDK Protocol（WebSocket）
//   不冲突 —— 两条独立路径。realtime_hub.ts 顶上"严禁 WebSocket"是
//   针对它自己那条路径（小程序 WS 老问题）；SDK Protocol Dev Plan 已
//   评估过：subprotocol biu-sdk.v1.json + 25s 心跳能克服小程序 6min idle
//   断线问题（详见 docs/BiuMind-Agent-Plane-WebSocket-Transport.md §14）。
//
// **跟 Flutter BiuClient 对应**：API 形态一致：
//   - connect(sessionId, sessionToken) → 立即返回，后台跑 socket 升级
//   - frames callback → 收解过的 SDK Protocol 帧
//   - sendUserText(text) → 简化封装；包成 type=user message
//   - close() → 主动断
//
// **为什么不引 vitest 单测**：miniapp 暂无测试框架；S10-2 一并加。
// 当前依赖 type-check + 实际小程序模拟器手测覆盖。

import { getAccessToken } from '../core/token_manager';

/// brain Agent Plane WS base URL；末尾不带 `/`，class 内部拼路径。
const BASE_URL =
  ((import.meta.env.VITE_BIU_AGENT_PLANE_URL as string) || '').replace(
    /\/$/,
    ''
  ) ||
  ((import.meta.env.VITE_BIU_API_BASE as string) || '');

/// SDK Protocol v1 帧最小类型集 —— 只包 BiuClient 自己用的字段。
/// 完整 schema 在 lib/biu-agent-schemas/（S1-2~S1-5 落地）；那时再切到完整
/// type，这里先用结构化 partial 避免循环依赖。
export interface SDKFrame {
  type: string;
  uuid?: string;
  session_id?: string;
  // 数据平面常见字段
  text?: string;
  subtype?: string;
  // 工具相关
  tool_use_id?: string;
  tool_name?: string;
  // 兜底 —— 让客户端拿原始 JSON 自己渲染
  [k: string]: unknown;
}

/// connect 时的可选参数。
export interface ConnectOpts {
  sessionId: string;
  sessionToken: string;
  /// 重连请求重放从 sinceSeq+1 开始的帧。无续传需求时省略（默认 0=
  /// 实时拉，不重放）。
  sinceSeq?: number;
}

export interface BiuClientOptions {
  /// brain WS base URL 覆盖；不传走环境变量默认。
  baseUrl?: string;
  /// 连续重连最多次数到达后放弃 frames 回调结束。默认 8 次。
  reconnectMaxAttempts?: number;
  /// 收到 1008 / 4xx 时调，期望返回新 session_token。null 时直接断。
  onTokenExpired?: () => Promise<string | null>;
  /// 测试注入 —— 替换 uni.connectSocket。生产留 undefined。
  connector?: (uri: string, protocols: string[]) => BiuTransport;
}

/// 下游帧回调。每条帧已经 JSON.parse 过。
export type FrameHandler = (frame: SDKFrame) => void;

/// uni.connectSocket 抽象 —— BiuClient 不直接 reach uni global，让测试
/// 注入 fake transport 不需要 stub 整个 uni 对象。
export interface BiuTransport {
  /// 注册帧回调（String / ArrayBuffer 都可能；BiuClient 内部 JSON.parse
  /// String 帧，二进制帧 drop）。
  onMessage(cb: (data: string | ArrayBuffer) => void): void;
  /// 远端断（包括小程序 6min idle 断线）。
  onClose(cb: (reason: string) => void): void;
  /// 升级 / 网络错。
  onError(cb: (err: unknown) => void): void;
  /// 发一帧 String（JSON.stringify 过）。返回 Promise resolve 即送出。
  send(data: string): Promise<void>;
  /// 主动关。
  close(): Promise<void>;
}

/// 默认 connector —— 包 uni.connectSocket。 subprotocol 用 'biu-sdk.v1.json'
/// 跟 brain ingress 协议谈判。
function defaultConnector(uri: string, protocols: string[]): BiuTransport {
  // uni 类型由 @dcloudio/uni-app 提供；avoid TS error 用 any 兜底。
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const u = (globalThis as any).uni;
  if (!u || typeof u.connectSocket !== 'function') {
    throw new Error('BiuClient: uni.connectSocket unavailable');
  }
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  const task: any = u.connectSocket({
    url: uri,
    protocols,
    // 小程序连接超时设短；25s 心跳保活由 BiuClient 自己处理
    timeout: 30000,
  });
  return {
    onMessage(cb) {
      task.onMessage((res: { data: string | ArrayBuffer }) => cb(res.data));
    },
    onClose(cb) {
      task.onClose((res: { reason: string }) => cb(res.reason || ''));
    },
    onError(cb) {
      task.onError((err: unknown) => cb(err));
    },
    send(data) {
      return new Promise((resolve, reject) => {
        task.send({
          data,
          success: () => resolve(),
          fail: (err: unknown) => reject(err),
        });
      });
    },
    close() {
      return new Promise((resolve) => {
        task.close({
          code: 1000,
          success: () => resolve(),
          fail: () => resolve(), // 忽略关闭失败 —— GC 即可
        });
      });
    },
  };
}

/// BiuClient —— 单条 brain Agent Plane WS session。
export class BiuClient {
  private readonly opts: Required<
    Omit<BiuClientOptions, 'baseUrl' | 'onTokenExpired' | 'connector'>
  > & {
    baseUrl: string;
    onTokenExpired?: () => Promise<string | null>;
    connector: (uri: string, protocols: string[]) => BiuTransport;
  };

  private readonly handlers = new Set<FrameHandler>();
  private transport: BiuTransport | null = null;
  private sessionId: string | null = null;
  private sessionToken: string | null = null;
  private closedByUser = false;
  private reconnectAttempt = 0;
  private reconnectTimer: ReturnType<typeof setTimeout> | null = null;
  private heartbeatTimer: ReturnType<typeof setInterval> | null = null;

  constructor(opts: BiuClientOptions = {}) {
    this.opts = {
      baseUrl: opts.baseUrl?.replace(/\/$/, '') || BASE_URL,
      reconnectMaxAttempts: opts.reconnectMaxAttempts ?? 8,
      onTokenExpired: opts.onTokenExpired,
      connector: opts.connector ?? defaultConnector,
    };
  }

  /// 注册帧回调。返回 unsubscribe。多 handler 支持 broadcast 模式。
  onFrame(cb: FrameHandler): () => void {
    this.handlers.add(cb);
    return () => this.handlers.delete(cb);
  }

  /// 是否当前在线（已升级 + 没主动关）。
  get isConnected(): boolean {
    return this.transport !== null && !this.closedByUser;
  }

  /// 跟一条 session 建 WS 连接。返回前不等 ready —— 内部异步。
  async connect(opts: ConnectOpts): Promise<void> {
    this.sessionId = opts.sessionId;
    this.sessionToken = opts.sessionToken;
    this.closedByUser = false;
    this.reconnectAttempt = 0;
    await this.doConnect(opts.sinceSeq);
  }

  /// 续 token 后让 client 用新 token 重连（如果当前断了）。
  async refreshToken(newToken: string): Promise<void> {
    this.sessionToken = newToken;
    if (!this.transport && !this.closedByUser) {
      await this.doConnect();
    }
  }

  /// 发一帧给 brain。未连或已关 → 抛错。
  async send(frame: SDKFrame): Promise<void> {
    const t = this.transport;
    if (!t) {
      throw new Error('BiuClient.send: not connected');
    }
    await t.send(JSON.stringify(frame));
  }

  /// 简化封装：发一条 user message（type=user，content 走 Anthropic
  /// Messages API content block 数组）。uuid 由调用方生成。
  async sendUserText(text: string, userMessageUuid: string): Promise<void> {
    await this.send({
      type: 'user',
      uuid: userMessageUuid,
      session_id: this.sessionId ?? '',
      message: {
        role: 'user',
        content: [{ type: 'text', text }],
      },
    });
  }

  /// 主动关 —— frames 不再收到新帧；不重连。
  async close(): Promise<void> {
    this.closedByUser = true;
    this.clearTimers();
    const t = this.transport;
    this.transport = null;
    if (t) await t.close();
    this.handlers.clear();
  }

  // ── Internals ──────────────────────────────────────────────

  private async doConnect(sinceSeq?: number): Promise<void> {
    if (this.closedByUser) return;
    const sid = this.sessionId;
    const tok = this.sessionToken;
    if (!sid || !tok) {
      throw new Error('BiuClient: connect() must run before doConnect');
    }
    const baseHttp = this.opts.baseUrl;
    // http(s) → ws(s)
    const baseWs = baseHttp.replace(/^http(s?):/, 'ws$1:');
    let uri = `${baseWs}/v1/agent/sessions/${sid}/stream?session_token=${encodeURIComponent(tok)}`;
    if (sinceSeq && sinceSeq > 0) uri += `&since_seq=${sinceSeq}`;
    try {
      const t = this.opts.connector(uri, ['biu-sdk.v1.json']);
      this.transport = t;
      this.reconnectAttempt = 0;
      t.onMessage((data) => this.onWireFrame(data));
      t.onClose(() => this.onSocketDone());
      t.onError(() => this.onSocketDone());
      this.startHeartbeat();
    } catch (e) {
      console.warn('BiuClient connect failed', e);
      this.scheduleReconnect();
    }
  }

  private onWireFrame(data: string | ArrayBuffer): void {
    if (typeof data !== 'string') {
      // 二进制 drop —— 协议层只发 JSON 字符串
      return;
    }
    let parsed: SDKFrame;
    try {
      parsed = JSON.parse(data) as SDKFrame;
    } catch (e) {
      console.warn('BiuClient bad frame', e, data.slice(0, 80));
      return;
    }
    for (const cb of this.handlers) {
      try {
        cb(parsed);
      } catch (e) {
        console.warn('BiuClient handler threw', e);
      }
    }
  }

  private onSocketDone(): void {
    this.transport = null;
    this.clearHeartbeat();
    if (this.closedByUser) return;
    this.scheduleReconnect();
  }

  private scheduleReconnect(): void {
    if (this.closedByUser) return;
    this.reconnectAttempt += 1;
    if (this.reconnectAttempt > this.opts.reconnectMaxAttempts) {
      console.warn('BiuClient: reconnect exhausted, giving up');
      this.closedByUser = true;
      this.handlers.clear();
      return;
    }
    const ms = this.backoffMs(this.reconnectAttempt);
    this.reconnectTimer = setTimeout(async () => {
      // 第 3 次开始尝试 token 刷新（前两次假设网络抖动，token 还有效）
      if (this.opts.onTokenExpired && this.reconnectAttempt > 2) {
        try {
          const fresh = await this.opts.onTokenExpired();
          if (fresh) this.sessionToken = fresh;
        } catch (e) {
          console.warn('BiuClient onTokenExpired threw', e);
        }
      }
      await this.doConnect();
    }, ms);
  }

  private backoffMs(attempt: number): number {
    // 1s / 2s / 4s / 8s / 16s / 30s（cap）...
    const ms = 1000 * Math.pow(2, Math.min(attempt - 1, 5));
    return Math.min(Math.max(ms, 1000), 30000);
  }

  /// 心跳 25s —— 小程序 6min idle 断线兜底。SDK Protocol 协议层走
  /// keep_alive lifecycle 帧（可被 brain ingress 静默吞）。
  private startHeartbeat(): void {
    this.clearHeartbeat();
    this.heartbeatTimer = setInterval(() => {
      this.send({ type: 'keep_alive', uuid: '', session_id: this.sessionId ?? '' }).catch(
        (e) => console.warn('BiuClient heartbeat failed', e)
      );
    }, 25_000);
  }

  private clearHeartbeat(): void {
    if (this.heartbeatTimer) {
      clearInterval(this.heartbeatTimer);
      this.heartbeatTimer = null;
    }
  }

  private clearTimers(): void {
    if (this.reconnectTimer) {
      clearTimeout(this.reconnectTimer);
      this.reconnectTimer = null;
    }
    this.clearHeartbeat();
  }
}

/// connectAuthenticated —— 简化入口；从 token_manager 拿当前 access
/// token（refresh 走 token_manager 内部），然后 connect。
///
/// 上层业务代码用这个，不直接构造 BiuClient —— 跟 realtime_hub 同款。
export async function connectBiuClient(opts: {
  sessionId: string;
  sessionToken: string;
  sinceSeq?: number;
  baseUrl?: string;
}): Promise<BiuClient> {
  // session_token 已经由 brain /v1/agent/sessions 颁发，不用再走 access_token；
  // 这里 import getAccessToken 只为 onTokenExpired 时拿新 access 触发 refresh。
  const client = new BiuClient({
    baseUrl: opts.baseUrl,
    onTokenExpired: async () => {
      // 兜底：期望上层 controller 先调 brain refresh-session-token 拿新
      // session_token，然后 client.refreshToken(newTok)。这里 access_token
      // 不直接当 session_token 用 —— 只是保证 token_manager 是热的。
      try {
        await getAccessToken();
      } catch {
        /* ignore */
      }
      return null;
    },
  });
  await client.connect({
    sessionId: opts.sessionId,
    sessionToken: opts.sessionToken,
    sinceSeq: opts.sinceSeq,
  });
  return client;
}
